tests/unit/test_remote_write.py

```diff
@@ -0,0 +1,263 @@
+
+    @patch.object(KubernetesServicePatch, "_service_object", new=lambda *args: None)
+    @patch.object(Prometheus, "reload_configuration", new=lambda _: True)
+    @patch("ops.testing._TestingPebbleClient.remove_path")
+    @patch("ops.testing._TestingPebbleClient.push", new=fake_push)
+    @patch("ops.testing._TestingModelBackend.network_get")
+    def test_endpoint_url_with_ingress_and_external_url(self, mock_net_get, *_):
+        ip = "1.1.1.1"
+        net_info = {"bind-addresses": [{"interface-name": "ens1", "addresses": [{"value": ip}]}]}
+        mock_net_get.return_value = net_info
+
+        self.harness.update_config({"web-external-url": "my_happy_ingress"})
+
+        ingress_rel_id = self.harness.add_relation("ingress", "nginx-ingress")
+        self.harness.add_relation_unit(ingress_rel_id, "ingress/0")
+
+        self.harness.begin_with_initial_hooks()
+
+        cons_rel_id = self.harness.add_relation(RELATION_NAME, "consumer")
+        self.harness.add_relation_unit(cons_rel_id, "consumer/0")
+        self.assertEqual(
+            self.harness.get_relation_data(cons_rel_id, self.harness.charm.unit.name),
+            {"remote_write": json.dumps({"url": "http://my_happy_ingress:9090/api/v1/write"})},
+        )
+
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus())
+
+    @patch.object(KubernetesServicePatch, "_service_object", new=lambda *args: None)
+    @patch.object(Prometheus, "reload_configuration", new=lambda _: True)
+    @patch("ops.testing._TestingPebbleClient.remove_path")
+    @patch("ops.testing._TestingPebbleClient.push", new=fake_push)
+    @patch("ops.testing._TestingModelBackend.network_get")
+    def test_multiple_units_with_ingress(self, mock_net_get, *_):
+        ip = "1.1.1.1"
+        net_info = {"bind-addresses": [{"interface-name": "ens1", "addresses": [{"value": ip}]}]}
+        mock_net_get.return_value = net_info
+
+        self.harness.update_config({"web-external-url": "my_happy_ingress"})
+
+        ingress_rel_id = self.harness.add_relation("ingress", "nginx-ingress")
+        self.harness.add_relation_unit(ingress_rel_id, "ingress/0")
+
+        self.harness.begin_with_initial_hooks()
+
+        peers_rel_id = self.harness.charm.model.get_relation("prometheus-peers").id
+        self.harness.add_relation_unit(peers_rel_id, "prometheus/1")
+
+        cons_rel_id = self.harness.add_relation(RELATION_NAME, "consumer")
+        self.harness.add_relation_unit(cons_rel_id, "consumer/0")
+        self.assertEqual(
+            self.harness.get_relation_data(cons_rel_id, self.harness.charm.unit.name),
+            {"remote_write": json.dumps({"url": "http://my_happy_ingress:9090/api/v1/write"})},
+        )
+
+        self.assertEqual(
+            self.harness.charm.unit.status,
+            BlockedStatus(
+                "invalid combination of 'ingress', 'receive-remote-write' relations and multiple units"
+            ),
```

Comment:
This test is very fragile. Make the error message a constant in the implementation and assert against that in the test to prevent it from breaking if we decide to change the phrasing.

Reply:
Solved in 9dbaf8e